### PR TITLE
[WFLY-11255] Managed Executor Runtime Stats

### PR DESCRIPTION
**Pull request by emmartins**
Issues: https://issues.jboss.org/browse/WFLY-11255 , https://issues.jboss.org/browse/WFLY-12467
WFLY-11255 Proposal PR: https://github.com/wildfly/wildfly-proposals/pull/243

### CODE REVIEW

**Comment by bstansberry** [comment]
The impl here seems fine.

**Comment by emmartins** [comment]
@bstansberry rebased with master, I believe all in EAP7-1175 is *now* ok for including this on WFLY19

**Review by jamezp — APPROVED** [review]
The two comments are minor and IMO shouldn't hold up the PR unless another review feels it should.

**Review by jamezp — CHANGES_REQUESTED** [review]
Actually it looks like the test may have issues and requesting changes is the only way I know to clear my approval.

That part of the test, the barrier part, does look a bit racy.

File: testsuite/integration/basic/src/test/java/org/jboss/as/test/integration/ee/concurrent/ManagedExecutorServiceMetricsTestCase.java [hunk 1]
```diff
@@ -0,0 +1,248 @@
...
+        });
+        Assert.assertTrue(readStatsAttribute(pathAddress, ManagedExecutorServiceMetricsAttributes.CURRENT_QUEUE_SIZE) == 1);
+        // resume tasks running and confirm expected idle stats
+        barrier2.await();
+        Assert.assertTrue(readStatsAttribute(pathAddress, ManagedExecutorServiceMetricsAttributes.ACTIVE_THREAD_COUNT) == 0);
+        Assert.assertTrue(readStatsAttribute(pathAddress, ManagedExecutorServiceMetricsAttributes.COMPLETED_TASK_COUNT) == 5);
+        Assert.assertTrue(readStatsAttribute(pathAddress, ManagedExecutorServiceMetricsAttributes.CURRENT_QUEUE_SIZE) == 0);
+        Assert.assertTrue(readStatsAttribute(pathAddress, ManagedExecutorServiceMetricsAttributes.MAX_THREAD_COUNT) == 2);
+        Assert.assertTrue(readStatsAttribute(pathAddress, ManagedExecutorServiceMetricsAttributes.TASK_COUNT) == 5);
+        Assert.assertTrue(readStatsAttribute(pathAddress, ManagedExecutorServiceMetricsAttributes.THREAD_COUNT) == 2);
```
Comment (jamezp) [inline testsuite/integration/basic/src/test/java/org/jboss/as/test/integration/ee/concurrent/ManagedExecutorServiceMetricsTestCase.java]: All these would be racy as we should only be checking once we no all 3 threads have completed.

File: ee/src/main/java/org/jboss/as/ee/concurrent/ManagedExecutorRuntimeStatsImpl.java [hunk 2]
```diff
@@ -0,0 +1,79 @@
...
+    }
+
+    @Override
+    public int getMaxThreadsCount() {
+        return threadPoolExecutor.getLargestPoolSize();
+    }
+
+    @Override
+    public int getHungThreadsCount() {
+        final Collection hungThreads = abstractManagedExecutorService.getHungThreads();
```
[the remark this answers is not in the document]
Comment (emmartins) [reply ee/src/main/java/org/jboss/as/ee/concurrent/ManagedExecutorRuntimeStatsImpl.java]: abstractManagedExecutorService.getHungThreads() always returns a null collection if there are no hung threads

File: ee/src/main/java/org/jboss/as/ee/concurrent/ManagedExecutorServiceImpl.java [hunk 3]
```diff
@@ -22,32 +22,38 @@
...
  */
 public class ManagedExecutorServiceImpl extends org.glassfish.enterprise.concurrent.ManagedExecutorServiceImpl {
 
     private final ControlPoint controlPoint;
+    private final ManagedExecutorRuntimeStats runtimeStats;
 
     public ManagedExecutorServiceImpl(String name, ManagedThreadFactoryImpl managedThreadFactory, long hungTaskThreshold, boolean longRunningTasks, int corePoolSize, int maxPoolSize, long keepAliveTime, TimeUnit keepAliveTimeUnit, long threadLifeTime, ContextServiceImpl contextService, RejectPolicy rejectPolicy, BlockingQueue<Runnable> queue, ControlPoint controlPoint) {
         super(name, managedThreadFactory, hungTaskThreshold, longRunningTasks, corePoolSize, maxPoolSize, keepAliveTime, keepAliveTimeUnit, threadLifeTime, contextService, rejectPolicy, queue);
         this.controlPoint = controlPoint;
+        this.runtimeStats = new ManagedExecutorRuntimeStatsImpl(this);
```
[the remark this answers is not in the document]
Comment (emmartins) [reply ee/src/main/java/org/jboss/as/ee/concurrent/ManagedExecutorServiceImpl.java]: The stats class also uses the executor to retrieve hung threads, honestly I would prefer to don't add comments here about what such class does internally with the executor.

File: testsuite/integration/basic/src/test/java/org/jboss/as/test/integration/ee/concurrent/ManagedExecutorServiceMetricsTestCase.java (hunk 1 above)
Comment (emmartins) [reply testsuite/integration/basic/src/test/java/org/jboss/as/test/integration/ee/concurrent/ManagedExecutorServiceMetricsTestCase.java]: You are correct, there was a race between tasks completing and getting stats. Honestly I don't recall why I used execute() in this final test, anyway now fixed by replacing execute() with submit(), and then using each future.get() to ensure all tasks complete before checking for idle stats.

**Comment by emmartins** [comment]
@jamezp seems all good now

**Review by jamezp — APPROVED** [review]

**Pull request merged**